src/interface/state_descriptor.cpp

```diff
@@ -0,0 +1,347 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "interface/metadata.hpp"
+#include "interface/state_descriptor.hpp"
+
+namespace parthenon {
+
+// Helper class for ResolvePackages
+struct DependencyTracker {
+  std::unordered_set<std::string> provided_vars;
+  std::unordered_set<std::string> depends_vars;
+
+  Dictionary<int> overridable_vars;
+  Dictionary<std::unordered_map<int, bool>> sparse_added_provides;
+  Dictionary<std::unordered_map<int, bool>> sparse_added_depends;
+  Dictionary<std::vector<Metadata>> overridable_meta;
+
+  template <typename AdderPrivate, typename AdderProvides>
+  void Sort(const std::string &package, const std::string &var, const Metadata &metadata,
+            const AdderPrivate &add_private, const AdderProvides &add_provides) {
+    auto dependency = metadata.Dependency();
+    bool is_sparse = metadata.IsSet(Metadata::Sparse);
+    int sparse_id = metadata.GetSparseId();
+    if (dependency == Metadata::Private) {
+      add_private(package, var, metadata);
+    } else if (dependency == Metadata::Provides) {
+      if (provided_vars.count(var) > 0) {
+        if (is_sparse) { // will simplify with dense on block
+          if (sparse_added_provides[var][sparse_id]) {
+            std::stringstream ss;
+            ss << "Sparse variable " << var << "_" << sparse_id
+               << " provided by multiple packages" << std::endl;
+            PARTHENON_THROW(ss);
+          }
+        } else {
+          PARTHENON_THROW("Variable " + var + " provided by multiple packages");
+        }
+      }
+      if (is_sparse) sparse_added_provides[var][sparse_id] = true;
+      provided_vars.insert(var);
+      add_provides(package, var, metadata);
+    } else if (dependency == Metadata::Requires) {
+      depends_vars.insert(var);
```

Comment:
when a package depends on a variable, does it not also depend on at least some properties of it's metadata.  for example, a package may require something like "velocity" with shape vector(1,3), but some package may provide "velocity" as a scalar.  That's a bit contrived, but I think you see the point.  Maybe this is captured in some other way I haven't seen yet...

Reply:
just another quick note on this.  this may be a place where some notion of needing "at least" these properties might be useful.  btw, this could certainly be for later.

Reply:
Yeah this is a good point and it's not currently captured... I don't think it would be hard to add something. But expressing it in the right way might be kind of tricky. I say we massage this MR and merge it in and attack that question a bit later.